File: backend/internal/serverboot/payment_wire_contract_test.go
package serverboot

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// paymentWireContractFixture is the SINGLE cross-language definition of the
// payment wire contract. It lives with the mobile test suite that consumes it
// and is read from here so the backend half proves the canonical handler still
// emits exactly those keys. There is deliberately no second copy of the shape.
//
// Dart half:
//
//	apps/mobile/test/domains/finance/transaction/payment/data/dto/payment_wire_contract_test.dart
const paymentWireContractFixture = "../../../apps/mobile/test/fixtures/payment_wire_contract.json"

type paymentWireBlock struct {
	Handler       string                     `json:"handler"`
	ForbiddenKeys []string                   `json:"forbidden_response_keys"`
	Response      map[string]json.RawMessage `json:"response"`
}

type paymentWireFixture struct {
	CreatePaymentRequest map[string]json.RawMessage `json:"create_payment_request"`
	CreatePayment        paymentWireBlock           `json:"create_payment"`
	ListPaymentMethods   paymentWireBlock           `json:"list_payment_methods"`
	GetPayment           paymentWireBlock           `json:"get_payment"`
}

func loadPaymentWireFixture(t *testing.T) paymentWireFixture {
	t.Helper()

	raw, err := os.ReadFile(filepath.FromSlash(paymentWireContractFixture))
	if err != nil {
		t.Fatalf("read payment wire fixture %s: %v", paymentWireContractFixture, err)
	}

	var fixture paymentWireFixture
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("decode payment wire fixture: %v", err)
	}
	if len(fixture.CreatePayment.Response) == 0 ||
		len(fixture.ListPaymentMethods.Response) == 0 ||
		len(fixture.GetPayment.Response) == 0 {
		t.Fatal("payment wire fixture must declare all three response blocks")
	}
	return fixture
}

// ginHKeysInFunc returns every top-level key of every `gin.H{...}` literal
// inside the function whose signature starts with signature.
//
// Only depth-1 keys are collected, so nested literals (e.g. the inline
// func() *string closures used for nullable response fields) cannot leak keys
// into the set. String literals are skipped while scanning, so braces inside
// strings never affect brace depth.
func ginHKeysInFunc(t *testing.T, src, signature string) map[string]bool {
	t.Helper()

	start := strings.Index(src, signature)
	if start < 0 {
		t.Fatalf("function %q not found in dependencies.go", signature)
	}
	body := src[start:]
	if next := strings.Index(body[len(signature):], "\nfunc "); next >= 0 {
		body = body[:len(signature)+next]
	}

	keys := make(map[string]bool)
	const marker = "gin.H{"

	for offset := 0; ; {
		idx := strings.Index(body[offset:], marker)
		if idx < 0 {
			break
		}
		i := offset + idx + len(marker)
		depth := 1

		for i < len(body) && depth > 0 {
			ch := body[i]
			switch {
			case ch == '"':
				// Read the whole string literal; capture it if it sits at
				// depth 1 and is immediately followed by a key separator.
				j := i + 1
				for j < len(body) {
					if body[j] == '\\' {
						j += 2
						continue
					}
					if body[j] == '"' {
						break
					}
					j++
				}
				if j >= len(body) {
					i = j
					break
				}
				literal := body[i+1 : j]
				rest := strings.TrimLeft(body[j+1:], " \t")
				if depth == 1 && strings.HasPrefix(rest, ":") {
					keys[literal] = true
				}
				i = j + 1
			case ch == '{':
				depth++
				i++
			case ch == '}':
				depth--
				i++
			default:
				i++
			}
		}
		offset = i
	}

	if len(keys) == 0 {
		t.Fatalf("no gin.H keys found in %q", signature)
	}
	return keys
}

// TestPaymentWireContract_BackendEmitsFixtureKeys proves the canonical payment
// handlers still emit every response key the shared fixture (and therefore the
// mobile parsers) relies on, and none of the keys that were purged.
//
// This fails if the backend renames payment_id, drops payment_url, or
// re-introduces a dropped/renamed key such as net_amount or coin_discount.
func TestPaymentWireContract_BackendEmitsFixtureKeys(t *testing.T) {
	fixture := loadPaymentWireFixture(t)

	src, err := os.ReadFile("dependencies.go")
	if err != nil {
		t.Fatalf("read dependencies.go: %v", err)
	}
	code := string(src)

	if len(fixture.CreatePaymentRequest) == 0 {
		t.Fatal("fixture must declare the canonical create-payment request body")
	}

	blocks := []struct {
		name string
		wire paymentWireBlock
	}{
		{"create_payment", fixture.CreatePayment},
		{"list_payment_methods", fixture.ListPaymentMethods},
		{"get_payment", fixture.GetPayment},
	}

	for _, block := range blocks {
		emitted := ginHKeysInFunc(t, code, block.wire.Handler)

		// Positive proof: every declared key must actually be emitted.
		declared := make([]string, 0, len(block.wire.Response))
		for key := range block.wire.Response {
			declared = append(declared, key)
		}
		for _, key := range declared {
			if !emitted[key] {
				t.Errorf("%s: fixture declares %q but %s never emits it (wire contract drift)",
					block.name, key, block.wire.Handler)
			}
		}

		// Negative proof: purged / stale keys must not come back.
		for _, forbidden := range block.wire.ForbiddenKeys {
			if emitted[forbidden] {
				t.Errorf("%s: %s must not emit %q (purged or renamed wire key)",
					block.name, block.wire.Handler, forbidden)
			}
		}
	}

	// The request body is backend-bound by CreatePaymentRequest; assert the
	// canonical keys are the ones the struct declares.
	for _, key := range []string{"order_id", "payment_method_code", "price_snapshot_id"} {
		if _, ok := fixture.CreatePaymentRequest[key]; !ok {
			t.Errorf("fixture request body is missing canonical key %q", key)
		}
	}
	if _, ok := fixture.CreatePaymentRequest["coin_discount"]; ok {
		t.Error("fixture request body must not declare the stale key coin_discount")
	}
	// CANONICAL PAY-B: K is fixed at Order creation (pricing_tokens.coins_used);
	// POST /payments must NOT accept a client coin authority.
	if _, ok := fixture.CreatePaymentRequest["coins_to_use"]; ok {
		t.Error("fixture request body must not declare coins_to_use — payment derives K from the pricing token")
	}

	for _, key := range []string{`json:"order_id"`, `json:"payment_method_code"`, `json:"price_snapshot_id"`} {
		if !strings.Contains(code, key) {
			t.Errorf("CreatePaymentRequest binding is missing %s", key)
		}
	}
	if strings.Contains(code, `json:"coin_discount"`) {
		t.Error("CreatePaymentRequest must not bind the outdated json key coin_discount")
	}
	if strings.Contains(code, `json:"coins_to_use"`) {
		t.Error("CreatePaymentRequest must not bind coins_to_use — payment derives K from the pricing token")
	}
}

// TestPaymentWireContract_NoDroppedMoneyColumnOnPaymentReads pins the reason
// net_amount must stay absent: migration 000037 dropped payments.net_amount, so
// no payment response may re-emit it.
func TestPaymentWireContract_NoDroppedMoneyColumnOnPaymentReads(t *testing.T) {
	src, err := os.ReadFile("dependencies.go")
	if err != nil {
		t.Fatalf("read dependencies.go: %v", err)
	}
	code := string(src)

	for _, signature := range []string{
		"func (h *CorePaymentHandler) CreatePayment(",
		"func (h *CorePaymentHandler) GetPayment(",
		"func (h *CorePaymentHandler) ListPaymentMethods(",
	} {
		emitted := ginHKeysInFunc(t, code, signature)
		if emitted["net_amount"] {
			t.Errorf("%s must not emit net_amount (payments.net_amount dropped in migration 000037)", signature)
		}
		if emitted["coin_discount"] {
			t.Errorf("%s must not emit coin_discount (renamed to coins_to_use in migration 000036)", signature)
		}
	}

	if !strings.Contains(code, `"coins_to_use"`) {
		t.Fatal("payment responses must expose the canonical coins_to_use key")
	}
}

File: apps/mobile/test/domains/finance/transaction/payment/data/dto/payment_wire_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:labuda/core/api/api_client.dart';
import 'package:labuda/core/common/types/payment_types.dart';
import 'package:labuda/domains/finance/transaction/payment/data/dto/payment_dto.dart';
import 'package:labuda/domains/finance/transaction/payment/data/remote/payment_remote_datasource.dart';
import 'package:labuda/domains/finance/transaction/payment/data/repositories/payment_repository_impl.dart';
import 'package:labuda/domains/finance/transaction/payment/domain/entities/payment.dart';
import 'package:labuda/shared/services/logger_service.dart';

/// CROSS-LANGUAGE PAYMENT WIRE CONTRACT PROOF (Dart half).
///
/// Single source of truth: test/fixtures/payment_wire_contract.json.
/// The Go half (backend/internal/serverboot/payment_wire_contract_test.go)
/// proves the canonical handler (CorePaymentHandler.CreatePayment /
/// GetPayment) emits exactly the keys declared in that fixture. This half
/// proves the mobile parsers consume exactly those keys — through the REAL
/// datasource and repository, not through a hand-rolled map.
///
/// This test fails if:
///   - the backend renames payment_id / drops payment_url (Go half fails, and
///     the fixture values below stop matching the real response),
///   - mobile re-introduces a required `net_amount` (parsing the fixture, which
///     has no such key, would throw),
///   - mobile expects the stale `id` / `amount` / `currency` on the create
///     response (parsing would throw / assert wrong values),
///   - the request body stops using the canonical `coins_to_use` key.
const String _fixturePath = 'test/fixtures/payment_wire_contract.json';

Map<String, dynamic> _fixture() =>
    jsonDecode(File(_fixturePath).readAsStringSync()) as Map<String, dynamic>;

Map<String, dynamic> _block(String name) =>
    _fixture()[name] as Map<String, dynamic>;

Map<String, dynamic> _response(String name) =>
    Map<String, dynamic>.from(_block(name)['response'] as Map<String, dynamic>);

List<String> _forbiddenKeys(String name) =>
    (_block(name)['forbidden_response_keys'] as List<dynamic>)
        .map((e) => e.toString())
        .toList();

/// Envelope-wrapped response, exactly as the backend `response.Success` helper
/// sends it (`{success, data, timestamp}`).
Map<String, dynamic> _envelope(Map<String, dynamic> data) => {
  'success': true,
  'data': data,
  'timestamp': '2026-09-18T10:00:00Z',
};

class _RecordingApiClient implements ApiClient {
  String? lastGetPath;
  String? lastPostPath;
  dynamic lastPostData;
  Map<String, dynamic>? lastGetQuery;
  int responseStatusCode = 200;
  dynamic getPayload = <String, dynamic>{'data': <String, dynamic>{}};
  dynamic postPayload = <String, dynamic>{'data': <String, dynamic>{}};

  @override
  Future<Response<T>> get<T>(
    String path, {
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
  }) async {
    lastGetPath = path;
    lastGetQuery = queryParameters;
    return Response<T>(
      requestOptions: RequestOptions(path: path),
      data: getPayload as T,
      statusCode: responseStatusCode,
    );
  }

  @override
  Future<Response<T>> post<T>(
    String path, {
    data,
    Map<String, dynamic>? queryParameters,
    Options? options,
    CancelToken? cancelToken,
  }) async {
    lastPostPath = path;
    lastPostData = data;
    return Response<T>(
      requestOptions: RequestOptions(path: path),
      data: postPayload as T,
      statusCode: responseStatusCode,
    );
  }

  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}

({PaymentRepositoryImpl repo, _RecordingApiClient client}) _repo() {
  final client = _RecordingApiClient();
  return (
    repo: PaymentRepositoryImpl(
      datasource: PaymentRemoteDatasource(client),
      logger: LoggerService.instance,
    ),
    client: client,
  );
}

void main() {
  group('payment wire contract fixture is single-source and alias-free', () {
    test('the canonical responses never carry obsolete keys', () {
      // Guards the fixture itself: it can never legitimize a dropped or
      // renamed wire key (net_amount was dropped by migration 000037).
      for (final block in ['create_payment', 'get_payment']) {
        final response = _response(block);
        for (final forbidden in _forbiddenKeys(block)) {
          expect(
            response.containsKey(forbidden),
            isFalse,
            reason: '$block response must not declare "$forbidden"',
          );
        }
      }
    });
  });

  group('POST /payments (create payment intent)', () {
    test('the canonical response parses into the payment intent', () async {
      final response = _response('create_payment');
      final r = _repo();
      r.client.postPayload = _envelope(response);

      final result = await r.repo.createPayment(
        const CreatePaymentRequest(
          orderId: '5c1e2f3a-4b5c-4d6e-8f90-1a2b3c4d5e6f',
          paymentMethodCode: 'bank_transfer',
        ),
      );

      expect(r.client.lastPostPath, '/payments');
      expect(result.isSuccess, isTrue, reason: result.failure?.message);

      final intent = result.data!;
      expect(intent.paymentId, response['payment_id']);
      expect(intent.paymentNumber, response['payment_number']);
      expect(intent.status, 'pending');
      expect(intent.paymentUrl, response['payment_url']);
      expect(intent.paymentMethodCode, 'bank_transfer');
      expect(intent.buyerPaymentFeeAmount, 4000);
      expect(intent.grossAmount, 114000);
      expect(intent.coinsToUse, 0);
      expect(intent.coinDiscountAmount, 0);
      expect(intent.referenceType, 'order');
      expect(intent.referenceId, response['reference_id']);
      expect(intent.expiredAt, DateTime.utc(2026, 9, 19, 10, 0, 0));

      // Checkout navigates to the in-app WebView only when a payment URL is
      // present and it is passed through unmodified.
      expect(intent.requiresRedirect, isTrue);
      expect(intent.paymentUrl, contains('midtrans.com'));
    });

    test('the request body uses the canonical create-payment keys', () async {
      final request = _block('create_payment_request');
      final expectedKeys = request.keys.toSet();
      final r = _repo();
      r.client.postPayload = _envelope(_response('create_payment'));

      await r.repo.createPayment(
        CreatePaymentRequest(
          orderId: request['order_id'] as String,
          paymentMethodCode: request['payment_method_code'] as String,
          priceSnapshotId: request['price_snapshot_id'] as String?,
        ),
      );

      final body = r.client.lastPostData as Map<String, dynamic>;
      expect(body.keys.toSet(), expectedKeys);
      expect(body, equals(request));
      // PAY-B: K is fixed at Order creation; the payment request must not carry
      // a client coin authority. The stale `coin_discount` key is never sent.
      expect(body.containsKey('coins_to_use'), isFalse);
      expect(body.containsKey('coin_discount'), isFalse);
    });

    test('a payload without the stale id/amount/currency keys still parses', () {
      final response = _response('create_payment');
      expect(response.containsKey('id'), isFalse);
      expect(response.containsKey('amount'), isFalse);
      expect(response.containsKey('currency'), isFalse);

      final dto = PaymentIntentDto.fromJson(response);
      expect(dto.paymentId, response['payment_id']);
      expect(dto.grossAmount, 114000);
    });

    test('unknown future response keys do not break parsing', () {
      final response = _response('create_payment')
        ..['some_future_field'] = 'ignored';
      final dto = PaymentIntentDto.fromJson(response);
      expect(dto.paymentId, response['payment_id']);
    });
  });

  group('GET /payments/:id (payment resource)', () {
    test('the canonical response parses into the payment entity', () async {
      final response = _response('get_payment');
      final r = _repo();
      r.client.getPayload = _envelope(response);

      final result = await r.repo.getPayment('9f6f0b0a-1d2e-4c3b-8a4f-2b7c1e5d6a90');

      expect(
        r.client.lastGetPath,
        '/payments/9f6f0b0a-1d2e-4c3b-8a4f-2b7c1e5d6a90',
      );
      expect(result.isSuccess, isTrue, reason: result.failure?.message);

      final payment = result.data!;
      expect(payment.id, response['id']);
      expect(payment.paymentNumber, response['payment_number']);
      expect(payment.userId, response['user_id']);
      expect(payment.grossAmount, 114000);
      expect(payment.coinsToUse, 0);
      expect(payment.coinDiscountAmount, 0);
      expect(payment.status, PaymentStatus.pending);
      expect(payment.referenceType, 'order');
      expect(payment.paymentUrl, response['payment_url']);
      expect(payment.expiredAt, DateTime.utc(2026, 9, 19, 10, 0, 0));
      expect(payment.midtransOrderId, response['midtrans_order_id']);
      expect(payment.paidAt, isNull);
    });

    test('parsing must not require the dropped net_amount key', () {
      final response = _response('get_payment');
      expect(response.containsKey('net_amount'), isFalse);

      final dto = PaymentDto.fromJson(response);
      expect(dto.grossAmount, 114000);
      // gross_amount is the canonical charged amount on this wire; the mobile
      // model has no net_amount surface at all.
      expect(dto.coinsToUse, 0);
    });

    test('unknown future response keys do not break parsing', () {
      final response = _response('get_payment')..['some_future_field'] = 1;
      final dto = PaymentDto.fromJson(response);
      expect(dto.id, response['id']);
    });
  });

  group('GET /payments/methods (fee disclosure)', () {
    test('per-method fee/total are consumed as backend-presented values', () async {
      final response = _response('list_payment_methods');
      final r = _repo();
      r.client.getPayload = _envelope(response);

      final result = await r.repo.getPaymentMethodOptions(
        response['order_id'] as String,
      );

      expect(r.client.lastGetPath, '/payments/methods');
      expect(r.client.lastGetQuery, {'order_id': response['order_id']});
      expect(result.isSuccess, isTrue, reason: result.failure?.message);

      final option = result.data!.single;
      final method =
          (response['methods'] as List<dynamic>).single
              as Map<String, dynamic>;
      expect(option.methodCode, method['method_code']);
      expect(option.displayName, method['display_name']);
      expect(option.buyerPaymentFeeAmount, method['buyer_payment_fee_amount']);
      expect(option.totalPayableAmount, method['total_payable_amount']);
      // PAY-B CLOSED: the backend no longer accepts a client `coins_to_use`
      // query input on this endpoint. The buyer fee is computed on
      // `cash_amount = base_amount − K`, where K is read from the pricing-token
      // snapshot written at Order creation (pricing_tokens.coins_used). The
      // client expresses coin intent only once, as the boolean `use_coins` on
      // POST /orders, and never manufactures a coin count downstream.
      //
      // This client therefore sends only the required `order_id`, and the guard
      // below pins that it can never re-acquire a payment-time coin authority.
      expect(r.client.lastGetQuery!.containsKey('coins_to_use'), isFalse);
    });
  });
}
